components/brave_wallet/browser/eth_abi_decoder.cc

```diff
@@ -0,0 +1,291 @@
+// followed by the actual content.
+//
+// The parsed bytearray is serialised as a hex string prefixed by "0x".
+bool GetBytesHexFromData(const std::string& input,
+                         std::string* arg,
+                         size_t offset) {
+  CHECK(arg);
+
+  size_t pointer;
+  if (!GetSizeFromData(input, &pointer, offset))
+    return false;
+
+  size_t bytes_len;
+  if (!GetSizeFromData(input, &bytes_len, pointer * 2))
+    return false;
+
+  *arg = "0x" + input.substr(pointer * 2 + 64, bytes_len * 2);
+  return true;
+}
+
+// GetAddressArrayFromData parses a calldata sequence to extract a dynamic
+// array of addresses at the specified offset using head-tail encoding
+// mechanism. The encoding is similar to bytes, with the first 32 bytes
+// representing the number of elements in the array, followed by each element.
+//
+// The parsed data is joined together into a hex string prefixed by "0x".
+bool GetAddressArrayFromData(const std::string& input,
+                             std::string* arg,
+                             size_t offset) {
+  CHECK(arg);
+
+  size_t pointer;
+  if (!GetSizeFromData(input, &pointer, offset))
+    return false;
+
+  size_t array_len;
+  if (!GetSizeFromData(input, &array_len, pointer * 2))
+    return false;
+
+  size_t array_offset = pointer * 2 + 64;
+  *arg = "0x";
+  for (size_t i = 0; i < array_len; i++) {
+    std::string value;
+    if (!GetAddressFromData(input, &value, array_offset))
+      return false;
+    *arg += value.substr(2);
+    array_offset += 64;
+  }
+
+  return true;
+}
+
+bool ABIDecodeInternal(const std::vector<std::string> types,
+                       const std::string& data,
+                       std::vector<std::string>* tx_params,
+                       std::vector<std::string>* tx_args) {
+  size_t offset = 0;
+  size_t calldata_tail = 0;
+
+  for (const auto& type : types) {
+    if ((data.length() - offset) < 64)
+      return false;
+
+    std::string value;
+    if (type == "address") {
+      if (!GetAddressFromData(data, &value, offset))
+        return false;
+    } else if (type == "uint256") {
+      if (!GetUint256HexFromData(data, &value, offset))
+        return false;
+    } else if (type == "bool") {
+      if (!GetBoolFromData(data, &value, offset))
+        return false;
+    } else if (type == "bytes") {
+      if (!GetBytesHexFromData(data, &value, offset))
+        return false;
+    } else if (type == "address[]") {
+      if (!GetAddressArrayFromData(data, &value, offset))
+        return false;
+    } else {
+      // For unknown/unsupported types, we only extract 32-bytes. In case of
+      // dynamic types, this value is a calldata reference.
+      value = data.substr(offset, 64);
+    }
+
+    // On encountering a dynamic type, we extract the reference to the start
+    // of the tail section of the calldata.
+    if ((type == "bytes" || type == "string" ||
+         type.substr(type.length() - 2) == "[]") &&
```

Comment:
This goes out of bounds if type == "x"

Reply:
Replaced with `base::EndsWith(type, "[]")`.